tests/main/views/test_buyers.py

```diff
@@ -3700,3 +3737,110 @@ def test_flash_message_shown_on_brief_cancel(self):
+
+    def setup_method(self, method):
+        super(TestAwardOrCancelBrief, self).setup_method(method)
+        self.data_api_client_patch = mock.patch('app.main.views.buyers.data_api_client', autospec=True)
+        self.data_api_client = self.data_api_client_patch.start()
+
+        self.data_api_client.get_framework.return_value = api_stubs.framework(
+            slug='digital-outcomes-and-specialists-2',
+            status='live',
+            lots=[
+                api_stubs.lot(slug='digital-outcomes', allows_brief=True),
+            ]
+        )
+        self.brief = api_stubs.brief(
+            user_id=123,
+            framework_slug='digital-outcomes-and-specialists-2',
+            lot_slug="digital-outcomes",
+            status='closed'
+        )['briefs']
+
+        self.data_api_client.get_brief.return_value = {"briefs": self.brief}
+        self.login_as_buyer()
+
+    def teardown_method(self, method):
+        self.data_api_client_patch.stop()
+        super(TestAwardOrCancelBrief, self).teardown_method(method)
+
+    def test_404_if_user_is_not_brief_owner(self):
+        self.data_api_client.get_brief.return_value['briefs']['users'][0]['id'] = 234
+
+        res = self.client.get(self.url.format(brief_id=self.brief["id"]))
+
+        assert res.status_code == 404
+
+    @pytest.mark.parametrize('status', ['withdrawn', 'draft', 'live', 'cancelled', 'unsuccessful', 'awarded'])
+    def test_404_if_brief_not_closed(self, status):
+        self.data_api_client.get_brief.return_value['briefs']['status'] = status
+
+        res = self.client.get(self.url.format(brief_id=self.brief["id"]))
+
+        assert res.status_code == 404
+
+    @pytest.mark.parametrize('framework_status', ['live', 'expired'])
+    def test_200_for_acceptable_framework_statuses(self, framework_status):
+        self.data_api_client.get_framework.return_value['frameworks']['status'] = framework_status
+
+        res = self.client.get(self.url.format(brief_id=123))
+
+        assert res.status_code == 200
+
+    def test_that_no_option_chosen_triggers_error(self):
+        res = self.client.post(self.url.format(brief_id=123))
+
+        document = html.fromstring(res.get_data(as_text=True))
+        validation_message = document.xpath('//span[@class="validation-message"]')[0].text_content()
+
+        assert res.status_code == 400
+        assert "You need to answer this question." in validation_message
+        assert self.data_api_client.cancel_brief.called is False
+
+    def test_yes_redirect(self):
+        self.login_as_buyer()
+        res = self.client.post(self.url.format(brief_id=self.brief['id']), data={'award_or_cancel_decision': 'yes'})
+
+        redirect_text = html.fromstring(res.get_data(as_text=True)).text_content().strip()
+        expected_url = (
+            '/buyers/frameworks/digital-outcomes-and-specialists-2/requirements/digital-outcomes/{}/award-contract'
+        ).format(self.brief['id'])
```

Comment:
I may have misunderstood this, but are we asserting here (and in the other redirect tests) that the redirect url appears somewhere in the page content of destination page? Does that then follow that we have actually been redirected to that url?

Elsewhere in this test module we're able to assert redirect locations like this, without needing to do any text wrangling:
```
res = self.client.post('/wherever')
assert res.status_code == 302
assert res.location == 'http://localhost/destination'
```

Reply:
Nice catch

Reply:
Done :+1: